Use mio to replace Epoll

uran0sH opened the pull request:
### Summary of the PR

Epoll is linux-specific. So we use mio, which is a cross-platform event notification, to replace Epoll.

### Requirements

Before submitting your PR, please make sure you addressed the following
requirements:

- [ x ] All commits in this PR have Signed-Off-By trailers (with
  `git commit -s`), and the commit message has max 60 characters for the
  summary and max 75 characters for each description line.
- [ x ] All added/changed functionality has a corresponding unit/integration
  test.
- [ x ] All added/changed public-facing functionality has entries in the "Upcoming 
  Release" section of CHANGELOG.md (if no such section exists, please create one).
- [ x ] Any newly added `unsafe` code is properly documented.

uran0sH committed:
Use mio to replace Epoll

Epoll is linux-specific. So we use mio, which is a cross-platform
event notification, to replace Epoll.

Signed-off-by: Wenyu Huang <huangwenyuu@outlook.com>

germag commented on vhost-user-backend/src/event_loop.rs:
```diff
@@ -3,61 +3,140 @@
+                .map(|interest| Interest::WRITABLE | interest)
+                .or(Some(Interest::WRITABLE));
+        }
+        interest.expect("Interest is invalid")
```
should be more like "Unknown EventSet", right?

uran0sH commented on vhost-user-backend/src/backend.rs:
```diff
@@ -144,7 +144,7 @@ pub trait VhostUserBackend: Send + Sync {
         &self,
         device_event: u16,
-        evset: EventSet,
+        evset: Interest,
```
Ok, changed it. And if we use `EventSet` in handle_event, I think we don't need `convert_event_to_interest`, so I removed it.

germag commented on vhost-user-backend/src/event_loop.rs:
```diff
@@ -3,52 +3,98 @@
+    }
+}
+
+impl From<EventSet> for Interest {
+    fn from(value: EventSet) -> Self {
+        let mut interest = None;
+        if value == EventSet::READABLE {
+            interest = interest
+                .map(|interest| Interest::READABLE | interest)
+                .or(Some(Interest::READABLE));
+        }
+        if value == EventSet::WRITABLE {
+            interest = interest
+                .map(|interest| Interest::WRITABLE | interest)
+                .or(Some(Interest::WRITABLE));
+        }
+        interest.expect("Unknown EventSet")
+    }
+}
```
As a second thought, I think this also must be a private function (like `to_interest(value: EventSet) -> Interest`), I don't want user to rely on this implementation

germag commented on vhost-user-backend/src/event_loop.rs:
```diff
@@ -3,52 +3,98 @@
-pub type VringEpollResult<T> = std::result::Result<T, VringEpollError>;
+pub type VringEpollResult<T> = std::result::Result<T, VringPollError>;
+
+bitflags! {
+    #[derive(Debug, PartialEq, PartialOrd)]
+    pub struct EventSet: u32 {
+        const READABLE = 1u32;
+        const WRITABLE = 2u32;
+    }
+}
```
the `bitflag!` has a "mismatch" with `mio::Interest` you cannot create an empty `mio::Interest` but you can do `EventSet::empty()` (that's the reason you need `.expect()` later). So I think it's better/simpler to just use an enum, like:

```rust
pub enum EventSet {
    Readable,
    Writable,
    All,
}
```

after this, I'm not sure if `EventSet` is the right name

germag commented on vhost-user-backend/src/event_loop.rs:
```diff
@@ -150,49 +218,30 @@ where
-                // handle_event() returns true if an event is received from the exit event fd.
-                if self.handle_event(ev_type, evset)? {
-                    break 'epoll;
+            for event in events.iter() {
```
nit: `for event in &events {`

germag commented on vhost-user-backend/src/event_loop.rs:
```diff
@@ -150,49 +218,30 @@ where
+                let token = event.token();
+
+                if let Some(evt_set) = event_to_event_set(event) {
+                    if self.handle_event(token.0 as u16, evt_set)? {
```
Since `token.0` is a `usize`, are we sure this cannot be truncated?

uran0sH replied:
In handle_event, the data is u16. But in register_event, the data is u64. Maybe we should all use u16 or u64?

stefano-garzarella replied:
> > nit: maybe we can move these changes in a separate commit (same PR, before adding `mio`)
> 
> It might not be moved to a separate commit. Because `Epoll::new(events: EventSet, data: u64)`, so we can't change `u64->u16` in `register_event` before using `mio`

why not? 
I mean, for sure we need other changes, but why it depends on `mio`?

uran0sH replied:
> > > nit: maybe we can move these changes in a separate commit (same PR, before adding `mio`)
> > 
> > 
> > It might not be moved to a separate commit. Because `Epoll::new(events: EventSet, data: u64)`, so we can't change `u64->u16` in `register_event` before using `mio`
> 
> why not? I mean, for sure we need other changes, but why it depends on `mio`?

Not depend on `mio`, just need some extra changes, like: `EpollEvent::new(ev_type, data as u64)`, `self.handle_event(ev_type, evset.try_into().unwarp())`

germag commented on vhost-user-backend/src/event_loop.rs:
```diff
@@ -135,13 +185,31 @@ where
+        self.registry
+            .register(
+                &mut SourceFd(&fd),
+                Token(data as usize),
```
since `data: u64`  and `pub struct Token(pub usize);` are we sure this cannot be truncated?

stefano-garzarella replied:
nit: maybe we can move these changes in a separate commit (same PR, before adding `mio`)

uran0sH committed:
Change u64 to usize in un/register_event

We can safely use usize instead of u64, because normally we will
not register a data that exceeds the size of usize

Signed-off-by: Wenyu Huang <huangwenyuu@outlook.com>

uran0sH committed:
Change device_event in handle_event() from u16 to usize

When register_event we pass in the value of u32/u64, so we don't
need to convert it to u16 and then pass it to handle_event.

Signed-off-by: Wenyu Huang <huangwenyuu@outlook.com>

stefano-garzarella commented on vhost-user-backend/src/event_loop.rs:
```diff
@@ -154,58 +219,39 @@ where
+            for event in &events {
+                let token = event.token();
+
+                if let Some(evt_set) = event_to_event_set(event) {
```
In the previous code we printed `epoll: ignoring unknown event set:...`, now we are just ignoring them, is it fine?

uran0sH replied:
Add it back again

stefano-garzarella commented on vhost-user-backend/CHANGELOG.md:
```diff
@@ -6,6 +6,7 @@
 ### Changed
 - [[#308](https://github.com/rust-vmm/vhost/pull/308)] Replace Eventfd with EventNotifier/EventConsumer.
 - [[#321](https://github.com/rust-vmm/vhost/pull/321)] Don't take ownership of listener in `VhostUserDaemon::start`.
+- [[316]](https://github.com/rust-vmm/vhost/pull/316) Use mio to replace Epoll. Expose event_loop::EventSet.
```
```suggestion
- [[316]](https://github.com/rust-vmm/vhost/pull/316)] Use mio to replace Epoll. Expose event_loop::EventSet.
```

stefano-garzarella commented on vhost-user-backend/src/event_loop.rs:
```diff
@@ -3,53 +3,90 @@
 
 /// Result of vring epoll operations.
-pub type VringEpollResult<T> = std::result::Result<T, VringEpollError>;
+pub type VringEpollResult<T> = std::result::Result<T, VringPollError>;
```
Should we rename this to `VringPollResult` ?

stefano-garzarella commented on vhost-user-backend/src/event_loop.rs:
```diff
@@ -58,7 +95,11 @@ pub type VringEpollResult<T> = std::result::Result<T, VringEpollError>;
 /// - remove registered file descriptors from the epoll fd
 /// - run the event loop to handle pending events on the epoll fd
 pub struct VringEpollHandler<T: VhostUserBackend> {
```
Ditto, should we rename this to `VringPollHandler`?

If it's a lot of renaming, I'd do that in a separate commit where we just rename structs (before or after using `mio`) to simplify the review.

stefano-garzarella commented:
@uran0sH thanks for restarting this work! Please, can you rebase this on current main.

FYI I'll be off from March 7 to 29 with limited internet access, so not sure I can review this before leaving, I'll do my best, but I'll leave to @germag and others handle it.

germag commented:
Hi, this LGTM

@name1e5s can you try to build it again?

eric-14 commented:
@uran0sH  building vhost backend on my macOs still fails due to eventfd in vhost/backend.rs. I am porting Vhostbackend trait methods to eventconsumer and eventNotifier. Are you considering taking this approach ?

uran0sH commented:
> @uran0sH building vhost backend on my macOs still fails due to eventfd in vhost/backend.rs. I am porting Vhostbackend trait methods to eventconsumer and eventNotifier. Are you considering taking this approach ?

Yeah, but the changes still are in my own repo: https://github.com/uran0sH/vhost/commit/54d673aa851f16c4828a6a661179de6c892a4ad8, not submitted to the upstream. If you want to test, you can use this branch: https://github.com/uran0sH/vhost/tree/macos. And I also change the vhost-device. These changes in https://github.com/uran0sH/vhost-device/pull/2

stefano-garzarella commented:
> > @uran0sH building vhost backend on my macOs still fails due to eventfd in vhost/backend.rs. I am porting Vhostbackend trait methods to eventconsumer and eventNotifier. Are you considering taking this approach ?
> 
> Yeah, but the changes still are in my own repo: [uran0sH@54d673a](https://github.com/uran0sH/vhost/commit/54d673aa851f16c4828a6a661179de6c892a4ad8), not submitted to the upstream. If you want to test, you can use this branch: https://github.com/uran0sH/vhost/tree/macos. And I also change the vhost-device. These changes in [uran0sH/vhost-device#2](https://github.com/uran0sH/vhost-device/pull/2)

@uran0sH what is missing to submit it here?

uran0sH commented:
> > > @uran0sH building vhost backend on my macOs still fails due to eventfd in vhost/backend.rs. I am porting Vhostbackend trait methods to eventconsumer and eventNotifier. Are you considering taking this approach ?
> > 
> > 
> > Yeah, but the changes still are in my own repo: [uran0sH@54d673a](https://github.com/uran0sH/vhost/commit/54d673aa851f16c4828a6a661179de6c892a4ad8), not submitted to the upstream. If you want to test, you can use this branch: https://github.com/uran0sH/vhost/tree/macos. And I also change the vhost-device. These changes in [uran0sH/vhost-device#2](https://github.com/uran0sH/vhost-device/pull/2)
> 
> @uran0sH what is missing to submit it here?

I think nothing major is missing. I am not sure whether [uran0sH@54d673a](https://github.com/uran0sH/vhost/commit/54d673aa851f16c4828a6a661179de6c892a4ad8) should be submitted in this pr or as a sperate one. What do you think?

stefano-garzarella commented:
> I think nothing major is missing. I am not sure whether [uran0sH@54d673a](https://github.com/uran0sH/vhost/commit/54d673aa851f16c4828a6a661179de6c892a4ad8) should be submitted in this pr or as a sperate one. What do you think?

I think better a separate PR if they are not depending on each other.

eric-14 commented:
> > > > @uran0sH building vhost backend on my macOs still fails due to eventfd in vhost/backend.rs. I am porting Vhostbackend trait methods to eventconsumer and eventNotifier. Are you considering taking this approach ?
> > > 
> > > 
> > > Yeah, but the changes still are in my own repo: [uran0sH@54d673a](https://github.com/uran0sH/vhost/commit/54d673aa851f16c4828a6a661179de6c892a4ad8), not submitted to the upstream. If you want to test, you can use this branch: https://github.com/uran0sH/vhost/tree/macos. And I also change the vhost-device. These changes in [uran0sH/vhost-device#2](https://github.com/uran0sH/vhost-device/pull/2)
> > 
> > 
> > @uran0sH what is missing to submit it here?
> 
> I think nothing major is missing. I am not sure whether [uran0sH@54d673a](https://github.com/uran0sH/vhost/commit/54d673aa851f16c4828a6a661179de6c892a4ad8) should be submitted in this pr or as a sperate one. What do you think?

@uran0sH I found the following issue with https://github.com/uran0sH/vhost/tree/macos
   
I have fixed it on my branch https://github.com/eric-14/rust_vhost_/tree/experimental_mac

stefano-garzarella commented:
@uran0sH talking with @eric-14 he would like to complete this PR since it's blocking some stuff. Is it okay for you if he will take your work (keeping your authorship of course) and complete it to upstream ASAP?

uran0sH commented:
> @uran0sH talking with @eric-14 he would like to complete this PR since it's blocking some stuff. Is it okay for you if he will take your work (keeping your authorship of course) and complete it to upstream ASAP?

It's okay for me.

germag commented:
@eric-14 Thanks!

After looking at it (honestly, just skimming) again, there are some minor issues that should be fixed and a couple not so minor:
- `poller: Mutex<Poll>`  - Can the mutex be removed (I know `VringPollHandler` needs sync maybe that's the problem we need to fix idk yet)?, epoll (and the mio impl) is designed to work with a single thread.
-  `println!("ignoring unknown event set:...)` -> ` log::warn!` (or `debug!`?)

Less important:
- ` register_/unregister_listener` api `data` is too general name, but we do: `data <= self.backend.num_queues()`
- `register_event()` does `...map_err(std::io::Error::other)?;`  while `unregister_event()... .map_err(|e| std::io::Error::other(format!("Failed to deregister fd {fd}: {e}")))?;`
- We need to check if nothing important is ignored inside `event_to_event_set()`, for example `evt.is_error()`.

eric-14 commented:
@germag Thank you for the latest review. I will make the necessary changes and ping you.

eric-14 commented:
@germag, I have completed the changes. Do you prefer I create a new PR or commit to this PR?

stefano-garzarella commented:
> @germag, I have completed the changes. Do you prefer I create a new PR or commit to this PR?

I'd suggest to start a new PR, I'm not sure if you have rights to push on this.
